Added ability to specify 'role' field for select speaker messages for Group Chats (Replaces PR #2167)

marklysze opened the pull request:
Note: This replaces #2167 due to it being based on an older version of main.

## Why are these changes needed?
As per feature request #1861 ("[Feature Request]: Allowing user to specify the "role" field for select speaker messages"), this PR includes changes to the GroupChat constructor to allow the user to specify the role for the select speaker messages. This addresses the issue of role names when using Mistral through Mistral.AI's API.

Changes:
- Added `role_for_select_speaker_messages` to `GroupChat` constructor in `groupchat.py`
- Added @ekzhu's Mistral.AI notebook to documentation under Non-OpenAI models section
- Added @ekzhu's updates to the Using Non-OpenAI about page
- Added a Tips for Non-OpenAI models page to documentation (should be a basis for future tips)
- Added a sub-section on the Conversation Patterns tutorial explaining the new attribute and provided an example

## Related issue number
Addresses #1861 

## Checks

- [X] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [ ] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [X] I've made sure all auto checks have passed.

ekzhu commented on website/docs/topics/non-openai-models/cloud-mistralai_files/figure-markdown_strict/cell-11-output-2.svg:
We don't need to check in these files as they will be automatically generated in CI

ekzhu commented on website/docs/tutorial/conversation-patterns.ipynb:
```diff
@@ -1183,6 +1183,29 @@
+     "As part of the Group chat process, a select speaker message is sent to the LLM\n",
+     "to determine the next speaker.\n",
+     "\n",
+     "Each message in the chat sequence has a `role` attribute that is typically `user`,\n",
```
When the select_speaker_method is set to auto, which is a default

ekzhu reviewed (changes requested):
Unit tests to check if the select speaker messages are indeed following the role specified

marklysze commented:
> Unit tests to check if the select speaker messages are indeed following the role specified

Okay, I'll add to tests

marklysze commented:
> Unit tests to check if the select speaker messages are indeed following the role specified

I've added a test for this. I've also updated groupchat.py (sorry @jackgerrits!) to include a check that `role_for_select_speaker_messages` isn't blank or None.

ekzhu commented on autogen/agentchat/groupchat.py:
```diff
@@ -74,6 +75,7 @@ def custom_speaker_selection_func(
     speaker_transitions_type: Literal["allowed", "disallowed", None] = None
     enable_clear_history: Optional[bool] = False
     send_introductions: bool = False
+    role_for_select_speaker_messages: Optional[str] = "system"
```
If None case is not handled then the type hint should be just str or even better, Literal["user", "system", "assistant"]

Same with several other parameters in the class. We can have a separate PR to fix those parameters cc @joshkyh

marklysze replied:
Sounds like a good idea to change the hint to Literal["user", "system", "assistant"] to assist developers.

I have a check for None or empty string in __post_init__:

```
        if self.role_for_select_speaker_messages is None or len(self.role_for_select_speaker_messages) == 0:
            raise ValueError("role_for_select_speaker_messages cannot be empty or None.")
```

Just confirming I'll leave this in? The hint doesn't enforce the value and it can still be None or empty.

ekzhu replied:
It's okay to do it in a next PR. I have created an issue.

sonichi commented:
Could you please fix the conflicts? Thanks.
